contracts/hackatom/src/contract.rs

```diff
@@ -232,6 +254,40 @@ fn query_other_balance<S: Storage, A: Api, Q: Querier>(
     Ok(AllBalanceResponse { amount })
 }
 
+fn query_recurse<S: Storage, A: Api, Q: Querier>(
+    deps: &Extern<S, A, Q>,
+    depth: u32,
+    work: u32,
+    contract: HumanAddr,
+) -> StdResult<RecurseResponse> {
+    // perform all hashes as requested
+    let mut hashed: Vec<u8> = contract.as_str().as_bytes().to_vec();
+    if work > 0 {
+        let mut step = Sha256::digest(&hashed);
+        for _ in 1..work {
+            step = Sha256::digest(&step);
+        }
+        hashed = step.to_vec();
+    }
```

Comment:
What about simplifying to this?

```rust
    let mut hashed: Vec<u8> = Vec::new();
    for run in 0..work {
        hashed = Sha256::digest(&run.to_be_bytes()).to_vec();
    }
```

Reply:
The compiler might simplify it to just the last item as we throw the rest away.

Recording on then ensures it does all work

Reply:
I did simplify it a fair bit using your `.to_vec()` trick